### PR TITLE
HPCC-18801 Add CLI options for listing and attaching found files.

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Used large sample XREF data from external system with copies of cluster layouts, to check funtionality.
List and attach single files and wildcards.

### CODE REVIEW

**Comment by jakesmith** [comment]
@AttilaVamos - please review

**Review by AttilaVamos — COMMENTED** [review]
@jakesmith some comments.

File: dali/dfuXRefLib/XRefFilesNode.cpp [hunk 1]
```diff
@@ -272,7 +276,7 @@ bool CXRefFilesNode::RemoveLogical(const char* LogicalName,IUserDescriptor* udes
     StringBuffer tmpbuf;
         
 
-    IPropertyTree* pLogicalFileNode =  getDataTree().getBranch(xpath.str());
+    IPropertyTree* pLogicalFileNode =  queryDataTree().getBranch(xpath.str());
     if (!pLogicalFileNode) {
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/XRefFilesNode.cpp]: my approach is usually to reformat any method I alter, that doesn't usually extend to v. simple global search and replace type changes like this one.

File: dali/dfuXRefLib/XRefFilesNode.cpp [hunk 2]
```diff
@@ -285,7 +289,7 @@ bool CXRefFilesNode::RemoveLogical(const char* LogicalName,IUserDescriptor* udes
         errstr.appendf("Logical file %s all parts exist (not lost?))",LogicalName);
         return false;
     }
-    if (!getDataTree().removeTree(pLogicalFileNode)) {                  
+    if (!queryDataTree().removeTree(pLogicalFileNode)) {
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/XRefFilesNode.cpp]: as per above, but might change

File: dali/dfuXRefLib/XRefFilesNode.hpp [hunk 3]
```diff
@@ -48,29 +49,30 @@ class CXRefFilesNode : implements IXRefFilesNode, public CSimpleInterface
     bool m_bChanged;
     IPropertyTree& m_baseTree;
     Owned<IPropertyTree> m_DataTree;
-    StringBuffer _data;
+    MemoryBuffer _data;
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/XRefFilesNode.hpp]: Not new, but I can change.

File: dali/dfuXRefLib/XRefFilesNode.hpp [hunk 4]
```diff
@@ -48,29 +49,30 @@ class CXRefFilesNode : implements IXRefFilesNode, public CSimpleInterface
...
     StringBuffer prefixName;
     StringAttr rootdir;
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/XRefFilesNode.hpp]: Not new/related to changes in this PR, could change.

File: dali/dfuXRefLib/XRefNodeManager.cpp [hunk 5]
```diff
@@ -107,6 +107,7 @@ CXRefNode::CXRefNode(const char* NodeName, IRemoteConnection *_conn)
     IPropertyTree* cluster_ptree = m_conn->queryRoot()->queryPropTree(xpath.str());
     m_XRefTree.set(cluster_ptree);
     m_XRefTree->getProp("@name",m_origName);
+    rootDir.set(m_XRefTree->queryProp("@rootdir"));
     //DBGLOG("returning from CXRefNode::CXRefNode(const char* NodeName)");
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/XRefNodeManager.cpp]: not new, but yes I could remove.

File: dali/dfuXRefLib/XRefNodeManager.hpp [hunk 6]
```diff
@@ -94,6 +95,7 @@ class CXRefNode : implements IXRefNode ,implements IXRefProgressCallback, public
     StringBuffer m_dataStr;
     StringBuffer m_ClusterName;
     Mutex commitMutex;
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/XRefNodeManager.hpp]: yes but old code, not planning on touching in this PR.

File: dali/dfuXRefLib/dfuxreflib.cpp [hunk 7]
```diff
@@ -2732,30 +2732,170 @@ IPropertyTree * runXRefCluster(const char *cluster,IXRefNode *nodeToUpdate)
...
+        case xrefUpdate:
+        {
+            if (flags & PMupdateeclwatch)
+            {
+                if (nclusters==1)
+                {
+                    const char *cluster = *clusters;
+                    CXRefNodeManager nodemanager;
+                    Owned<IPropertyTree> tree = runXRef(nclusters,clusters,NULL,numthreads);
+                    if (tree) {
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfuXRefLib/dfuxreflib.cpp]: will change.

**Comment by jakesmith** [comment]
I will address the issues that changes in this PR involve, but probably not most of the ones that are unrelated/old e.g. untouched code with naming convention issues.

**Comment by jakesmith** [comment]
@AttilaVamos - made some formatting changes and changed getData() to more appropriately named queryData(). Have left member variables etc. as they were though, as not new and there are 100's of naming convention issues in all these files, not a change for this PR.

**Comment by AttilaVamos** [comment]
@jakesmith it seem good.

**Comment by jakesmith** [comment]
@richardkchapman - please merge

**Pull request merged**